tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py

```diff
@@ -806,3 +873,22 @@ def _tensor_gather_helper(gather_indices, gather_from, batch_size,
     output = array_ops.reshape(output, final_shape, name="output")
     output.set_shape(final_static_shape)
     return output
+
+
+def _maybe_sort_array_beams(t, parent_ids, sequence_length):
+  """Maybe sorts beams within a `TensorArray`.
+
+  Args:
+    t: A `TensorArray` of size `max_time` that contains `Tensor`s of shape
+      `[batch_size, beam_width, depth]`.
+    parent_ids: The parent ids of shape `[max_time, batch_size, beam_width]`.
+    sequence_length: The sequence length of shape `[batch_size, beam_width]`.
+
+  Returns:
+    A `TensorArray` where beams are sorted in each `Tensor` or `t` itself if it
+    is not a `TensorArray`.
+  """
+  if isinstance(t, tensor_array_ops.TensorArray):
```

Comment:
There's no guarantee that the TensorArray has shapes that are amenable to sorting.  what if it contains scalars?  what if the the shape is not `[batch_size, beam_width, depth]` but instead someone is storing previous time steps' state having shape `[?, depth]` where `?` changes from element to element?  This is perfectly valid - you can't stack this TensorArray but you *can* `concat` it.

In other words, this function is making some serious assumptions about the fact that only an `AttentionWrapper` might create TensorArrays.  Guard against it; enable by default, but if there's a failure raise a warning message that the user should disable the TensorArray sorting.

Additional checks you should use:

1. Only perform this if `t` has `infer_shape=True`.
2. Only perform this if `t.element_shape` is rank 3 and element_shape[2:3] == [beam_width, depth].

Reply:
Currently, the assumptions are:

1. the TensorArrays size is the time dimension
2. the TensorArrays contain Tensors with the same shape assumption as Tensors that are directly in the cell state

> Guard against it; enable by default, but if there's a failure raise a warning message that the user should disable the TensorArray sorting.

Do you mean to try/catch the call to `gather_tree_from_array` and expose a flag to disable any post process on TensorArrays?